{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Context for this discussion (also see [#19](https://github.com/hammerlab/pgv/issues/19#issuecomment-180023783)):\n",
    "\n",
    "```\n",
    "alex [5:54 PM] \n",
    "Hey Arman, I talked with Ryan for a while today about allele specific transcript selection. He’s going to try find variant reads with an FM index query. Unclear how to reconcile the variant reads once they’re all gathered.\n",
    "\n",
    "Anyway, it might be nice to benchmark against a dumb baseline. If you have time, what do you think of putting together something that just takes aligned RNA, finds all the reads containing a variant (probably using varlens) and then just the most common nucleotide for each flanking position?\n",
    "\n",
    "arman [5:55 PM] \n",
    "sure, sounds easy enough\n",
    "\n",
    "[5:55] \n",
    "so the input files will be a VCF file and a BAM file and we will output some summary stats on each variant if I understand you correctly\n",
    "\n",
    "[5:56] \n",
    "but what do you exactly mean by a flanking position? The bases next to the variant?\n",
    "\n",
    "alex [5:57 PM] \n",
    "Yeah. If you center the variant nucleotides, and then count up the {A,C,T,G} content at every position -1,-2,-3,&c to the left and +1,+2,+3, &c to the right (and normalize by number of reads containing those positions), can we just take the most common?\n",
    "\n",
    "[5:58] \n",
    "The output could be something like:\n",
    "\n",
    "-5 -4 -3 -2 -1 _ _ +1 +2 +3 +4 +5\n",
    "A 0.1 0.2 \n",
    "C 0.9\n",
    "T\n",
    "G\n",
    "(edited)\n",
    "\n",
    "[5:59] \n",
    "Oh jeez, that’s getting tedious to fill out\n",
    "\n",
    "arman [5:59 PM] \n",
    "yeah, sure -- something like a PSM representing a motif\n",
    "\n",
    "alex [5:59 PM] \n",
    "But you’d have independent probability distributions over {A,G,T,G} at each position before and after the variant nucleotides\n",
    "\n",
    "arman [6:00 PM] \n",
    "happy to do that. I will try to get something going during the weekend but the worst case scenerio, we will have it by the end of Monday.\n",
    "\n",
    "alex [6:00 PM] \n",
    "Extending out to something like 75bp in both directions (since 25mer vaccine peptide = 75bp)\n",
    "\n",
    "[6:00] \n",
    "The variant nucleotides should have no uncertainty in them since we filter the reads to definitely contain the variant\n",
    "\n",
    "[6:01] \n",
    "And the best estimate of the sequence would be just taking the most probable nucleotide at each position\n",
    "\n",
    "[6:01] \n",
    "It’s a bit naive since it treats positions independently, but if the entropy is low enough (i.e. not much splicing diversity) then could be OK\n",
    "\n",
    "[6:02] \n",
    "And relying on the alignments is, one the one hand making it harder to detect weird splices, but on the other hand could be safer than reference-free fishing for matching sequences.\n",
    "\n",
    "arman [6:03 PM] \n",
    "I agree. I think this would be great start and there is always the option to extend what we have from this and add some experimental stuff on top of it.\n",
    "\n",
    "[6:03] \n",
    "Cool, sounds like a great plan.\n",
    "\n",
    "alex [6:05 PM] \n",
    "Awesome\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def contexify(samfile, chromosome, location, allele, radius):\n",
    "    # This will be our score board\n",
    "    counts = np.zeros(shape=((radius * 2) + 1, 5)) # 5 slots for each of the bases\n",
    "    d = pd.DataFrame(counts,\n",
    "                     index=range(location - radius, location + radius + 1),\n",
    "                     columns=['A', 'C', 'G', 'T', 'N'])\n",
    "\n",
    "    # Let pysam pileup the reads covering our location of interest for us\n",
    "    for column in samfile.pileup(chromosome, location, location + 1):\n",
    "        # By default, our region is bigger than we want,\n",
    "        # so skip the other columns for now\n",
    "        if column.pos != location:\n",
    "            continue\n",
    "\n",
    "        # Iterate over reads covering our location\n",
    "        for read in column.pileups:\n",
    "            if not read.is_del and not read.is_refskip:  # clean up\n",
    "                pos = read.query_position  # relative location\n",
    "                seq = read.alignment.query_sequence  # read\n",
    "\n",
    "                # Filter reads that do not support the variant\n",
    "                if seq[pos] is not allele:\n",
    "                    continue\n",
    "\n",
    "                # Cursor is compatible with our score board indexing\n",
    "                cursor = location - pos - 1\n",
    "                for base in seq:  # Move along the sequence one base at a time\n",
    "                    cursor = cursor + 1\n",
    "\n",
    "                    # If the region is beyond our window, discard the data\n",
    "                    if (cursor < location - radius) or (cursor > location + radius):\n",
    "                        continue\n",
    "\n",
    "                    # Count++ within our scoreboard\n",
    "                    count = d[base][cursor]\n",
    "                    d[base][cursor] = count + 1\n",
    "\n",
    "    return d.transpose()  # Transpose it to make it look more natural"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inspired by [varlens](https://github.com/hammerlab/varlens#varlens-allele-support) examples, here is how this simple function works:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import urllib\n",
    "\n",
    "urltemplate = \"https://raw.githubusercontent.com/hammerlab/varlens/master/test/data/CELSR1/bams/{}\"\n",
    "url = urllib.URLopener()\n",
    "url.retrieve(urltemplate.format(\"bam_5.bam\"), \"bam_5.bam\")\n",
    "url.retrieve(urltemplate.format(\"bam_5.bam.bai\"), \"bam_5.bam.bai\")\n",
    "\n",
    "samfile = pysam.AlignmentFile(\"bam_5.bam\", \"rb\")\n",
    "\n",
    "# C -> T variant at this particular locus\n",
    "chromosome = \"chr22\"\n",
    "location = 46930258\n",
    "radius = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare the contexts for the variant and the reference alleles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>46930253</th>\n",
       "      <th>46930254</th>\n",
       "      <th>46930255</th>\n",
       "      <th>46930256</th>\n",
       "      <th>46930257</th>\n",
       "      <th>46930258</th>\n",
       "      <th>46930259</th>\n",
       "      <th>46930260</th>\n",
       "      <th>46930261</th>\n",
       "      <th>46930262</th>\n",
       "      <th>46930263</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   46930253  46930254  46930255  46930256  46930257  46930258  46930259  \\\n",
       "A         1         0         0         0         0         0         0   \n",
       "C         0         0         1         0         1         0         0   \n",
       "G         0         0         0         1         0         0         1   \n",
       "T         0         1         0         0         0         1         0   \n",
       "N         0         0         0         0         0         0         0   \n",
       "\n",
       "   46930260  46930261  46930262  46930263  \n",
       "A         0         0         0         0  \n",
       "C         0         1         0         0  \n",
       "G         0         0         1         0  \n",
       "T         1         0         0         1  \n",
       "N         0         0         0         0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele1 = \"T\"\n",
    "contexify(samfile, chromosome, location, allele1, radius)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>46930253</th>\n",
       "      <th>46930254</th>\n",
       "      <th>46930255</th>\n",
       "      <th>46930256</th>\n",
       "      <th>46930257</th>\n",
       "      <th>46930258</th>\n",
       "      <th>46930259</th>\n",
       "      <th>46930260</th>\n",
       "      <th>46930261</th>\n",
       "      <th>46930262</th>\n",
       "      <th>46930263</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1018</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1038</td>\n",
       "      <td>0</td>\n",
       "      <td>1079</td>\n",
       "      <td>1091</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1063</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1081</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1035</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>2</td>\n",
       "      <td>1025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1071</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   46930253  46930254  46930255  46930256  46930257  46930258  46930259  \\\n",
       "A      1018         0         0         0         0         0         0   \n",
       "C         0         1      1038         0      1079      1091         0   \n",
       "G         1         0         0      1061         0         0      1081   \n",
       "T         2      1025         0         0         0         0         1   \n",
       "N         1         2         0         0         0         0         0   \n",
       "\n",
       "   46930260  46930261  46930262  46930263  \n",
       "A         0         0         0         1  \n",
       "C         0      1063         0         0  \n",
       "G         1         0      1035         0  \n",
       "T      1071         0         0      1024  \n",
       "N         0         0         0         0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele2 = \"C\"\n",
    "contexify(samfile, chromosome, location, allele2, radius)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
